[css-anchor-position-1] Style resolution should visit children of anchor-positioned elements after it's resolved

tuankiet65 opened the pull request:
#### 8c48c1a274e9c031298fce6c195974f85c056c9d
<pre>
[css-anchor-position-1] Style resolution should visit children of anchor-positioned elements after it&apos;s resolved
<a href="https://rdar.apple.com/148608785">rdar://148608785</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291065">https://bugs.webkit.org/show_bug.cgi?id=291065</a>

Reviewed by Antti Koivisto.

There&apos;s an optimization in Style::TreeResolver::resolveComposedTree that
defers visiting children of a query container or anchor-positioned element
until it can be resolved (e.g an anchor-positioned element waiting for its
anchor dependencies to be laid out):
* On initial style resolution, save the element change and which descendants
  to resolve (the &quot;state&quot;), and skip visiting the children
* On subsequent style resolutions, restore the state and continue as usual.
  The code gets &quot;tricked&quot; into resuming visiting into its children, where
  the previous resolution left off.

This is implemented for query containers, but not for anchor-positioned
elements. This leads to a subtle bug: if an anchor-positioned element is
initially hidden (`display: none`, `visibility: hidden` or hidden popover),
and then shown (`display` other than none, `visibility: visible`, or shown
popover), style resolution will **not** visit its children:
* On initial style resolution after the visibility change, the code indicates all
  descendents must be visited. Since the parent can&apos;t be resolved until subsequent
  resolutions, we skip visiting the children but **doesn&apos;t** save the state.
* On subsequent style resolutions, the parent can be resolved, so we can resume
  visiting the children. But we don&apos;t restore the state to &quot;trick&quot; the code into
  thinking it has to do so, so it skips visiting the children.

The results is the style resolution/layout/render code never sees the children,
and therefore won&apos;t layout/render them.

The solution is to implement the state save and restore for anchor-positioned
elements. However, this logic is currently tightly coupled with query
containers, so make it generic so it can be use for both purposes.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-visibility-change-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-visibility-change-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-visibility-change.html:
    - Add cases where the anchor-center element is hidden using
      `visibility: hidden` or popover, then shown.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-visibility-change-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-visibility-change.html: Added.
    - Add test that checks that an initially hidden anchor-positioned element
      using `position-area` properly renders when shown.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/reference/position-area-visibility-change-ref.html: Added.
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveComposedTree):
    - Before checking whether to visit children or not, call
      resumeDescendantResolutionIfNeeded.
    - If deferring descendant resolution because of query containers or
      anchor-positioned elements, call deferDescendantResolution.

(WebCore::Style::TreeResolver::deferDescendantResolution):
    - New function that saves the state necessary to defer descendant resolution.

(WebCore::Style::TreeResolver::resumeDescendantResolutionIfNeeded):
    - New function that restores the state necessary to resume descendant resolution.

(WebCore::Style::TreeResolver::updateStateForQueryContainer):
    - Move the state saving and restore logic into separate methods
      (deferDescendantResolution, resumeDescendantResolutionIfNeeded)

(WebCore::Style::TreeResolver::resolve):
* Source/WebCore/style/StyleTreeResolver.h:
    - Move states relevant to descendant resolution deferral in QueryContainerState
      into DeferredDescendantResolutionState.

Canonical link: <a href="https://commits.webkit.org/294924@main">https://commits.webkit.org/294924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9b20fcc5040f43fe12707918e5510a96ca98117

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103348 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13345 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108521 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53991 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105387 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78540 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35490 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106354 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93233 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58873 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18001 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11274 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53348 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87752 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110898 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30486 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22509 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87537 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30851 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89430 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87174 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22224 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32037 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9757 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24820 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30415 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35780 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30223 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33548 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1233,30 +1236,45 @@ const RenderStyle* TreeResolver::existingStyle(const Element& element)
+    if (it == m_styleLayoutInterleavedElementStates.end())
+        return;
+
+    auto& state = it->value;
```
perhaps could be `const`

tuankiet65 replied:
Fixed.

Scony commented on Source/WebCore/style/StyleTreeResolver.h:
```diff
@@ -160,22 +163,31 @@ class TreeResolver {
+    // container queries), the first style pass will defer resolving descendants.
+    // This is resumed in subsequent style passes, when there are enough information
+    // to do so. This struct preserves internal resolution state when descendant
+    // resolution is deferred, so it could be restores when descendant resolution
```
```suggestion
    // resolution is deferred, so it could be restored when descendant resolution
```

tuankiet65 replied:
Fixed.

Scony reviewed (commented):
LGTM.

I very much like the changes in WPT tests as those cover more corner-cases. It would be nice to export those changes to WPT in the future.

anttijk commented on Source/WebCore/style/StyleTreeResolver.h:
```diff
@@ -160,22 +163,31 @@ class TreeResolver {
+    UncheckedKeyHashMap<Ref<Element>, StyleLayoutInterleavedElementState> m_styleLayoutInterleavedElementStates;
     bool m_needsInterleavedLayout { false };
 
+    struct QueryContainerState {
+        bool invalidated { false };
+    };
+    UncheckedKeyHashMap<Ref<Element>, QueryContainerState> m_queryContainerStates;
```
Maybe there is a way to merge this state too with the interleaved state in future.

anttijk commented on Source/WebCore/style/StyleTreeResolver.h:
```diff
@@ -160,22 +163,31 @@ class TreeResolver {
+    // to do so. This struct preserves internal resolution state when descendant
+    // resolution is deferred, so it could be restores when descendant resolution
+    // resumes.
+    struct StyleLayoutInterleavedElementState {
```
I would leave out the word Style here as we are in Style namespace.

tuankiet65 replied:
I changed it to `DeferredDescendantResolutionState` which is a bit more relevant

anttijk reviewed (approved):
Looks good

tuankiet65 committed:
[css-anchor-position-1] Style resolution should visit children of anchor-positioned elements after it's resolved
rdar://148608785
https://bugs.webkit.org/show_bug.cgi?id=291065

Reviewed by Antti Koivisto.

There's an optimization in Style::TreeResolver::resolveComposedTree that
defers visiting children of a query container or anchor-positioned element
until it can be resolved (e.g an anchor-positioned element waiting for its
anchor dependencies to be laid out):
* On initial style resolution, save the element change and which descendants
  to resolve (the "state"), and skip visiting the children
* On subsequent style resolutions, restore the state and continue as usual.
  The code gets "tricked" into resuming visiting into its children, where
  the previous resolution left off.

This is implemented for query containers, but not for anchor-positioned
elements. This leads to a subtle bug: if an anchor-positioned element is
initially hidden (`display: none`, `visibility: hidden` or hidden popover),
and then shown (`display` other than none, `visibility: visible`, or shown
popover), style resolution will **not** visit its children:
* On initial style resolution after the visibility change, the code indicates all
  descendents must be visited. Since the parent can't be resolved until subsequent
  resolutions, we skip visiting the children but **doesn't** save the state.
* On subsequent style resolutions, the parent can be resolved, so we can resume
  visiting the children. But we don't restore the state to "trick" the code into
  thinking it has to do so, so it skips visiting the children.

The results is the style resolution/layout/render code never sees the children,
and therefore won't layout/render them.

The solution is to implement the state save and restore for anchor-positioned
elements. However, this logic is currently tightly coupled with query
containers, so make it generic so it can be use for both purposes.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-visibility-change-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-visibility-change-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-center-visibility-change.html:
    - Add cases where the anchor-center element is hidden using
      `visibility: hidden` or popover, then shown.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-visibility-change-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-visibility-change.html: Added.
    - Add test that checks that an initially hidden anchor-positioned element
      using `position-area` properly renders when shown.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/reference/position-area-visibility-change-ref.html: Added.
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveComposedTree):
    - Before checking whether to visit children or not, call
      resumeDescendantResolutionIfNeeded.
    - If deferring descendant resolution because of query containers or
      anchor-positioned elements, call deferDescendantResolution.

(WebCore::Style::TreeResolver::deferDescendantResolution):
    - New function that saves the state necessary to defer descendant resolution.

(WebCore::Style::TreeResolver::resumeDescendantResolutionIfNeeded):
    - New function that restores the state necessary to resume descendant resolution.

(WebCore::Style::TreeResolver::updateStateForQueryContainer):
    - Move the state saving and restore logic into separate methods
      (deferDescendantResolution, resumeDescendantResolutionIfNeeded)

(WebCore::Style::TreeResolver::resolve):
* Source/WebCore/style/StyleTreeResolver.h:
    - Move states relevant to descendant resolution deferral in QueryContainerState
      into DeferredDescendantResolutionState.

Canonical link: https://commits.webkit.org/294924@main
